File: backend/internal/scoring/engine.go
// Package scoring implements the PlayArena live scoring engine.
//
// The engine is stateless and has no database access or side effects.
// All score derivation is pure computation over the effective event log
// supplied by the caller.  match_events is the authoritative source of truth;
// this package only reads, never writes.
//
// Kabaddi scoring rules applied by Compute:
//
//	raid_successful    payload.points      → raiding team
//	bonus_point_awarded +1                → raiding team
//	tackle_successful   +1                → defending team
//	super_tackle        +2                → defending team
//	all_out             payload.bonus_pts → opponent of payload.team_id
//	penalty_awarded    payload.points      → attributed team
//	super_raid          0 (analytics only — raid_successful carries the points)
//	all other types     0
package scoring

import (
	db "github.com/4yushraman-jpg/playarena/db/sqlc"
	"github.com/4yushraman-jpg/playarena/internal/platform/pgutil"
)

// ScoreEngine derives match scores from an effective event log.
// It is stateless: construct once and reuse freely across goroutines.
type ScoreEngine struct{}

// NewScoreEngine returns a ScoreEngine ready for use.
func NewScoreEngine() *ScoreEngine { return &ScoreEngine{} }

// Compute calculates home and away scores from the provided events.
//
// The caller must pass only effective (non-cancelled) events.  Filtering is
// the repository's responsibility (GetEffectiveMatchEventsForScore query);
// the engine does not re-apply the cancellation filter.
//
// score_correction events present in the slice contribute zero points —
// they exist only to carry the cancels_event_id reference and are already
// excluded from the effective log on the target side.
//
// No data is persisted.  Call this function as often as needed.
func (e *ScoreEngine) Compute(match *db.Match, events []db.MatchEvent) ScoreResult {
	res := ScoreResult{
		MatchID:     pgutil.UUIDToString(match.ID),
		MatchStatus: string(match.Status),
		IsWalkover:  match.IsWalkover,
	}

	if match.HomeTeamID.Valid {
		res.HomeTeamID = pgutil.UUIDToString(match.HomeTeamID)
		res.AwayTeamID = pgutil.UUIDToString(match.AwayTeamID)
	} else {
		res.HomePlayerID = pgutil.UUIDToString(match.HomePlayerID)
		res.AwayPlayerID = pgutil.UUIDToString(match.AwayPlayerID)
	}

	for i := range events {
		pts, s := e.pointsForEvent(&events[i], match)
		switch s {
		case sideHome:
			res.HomeScore += pts
		case sideAway:
			res.AwayScore += pts
		}
	}
	return res
}

// pointsForEvent returns the point value and recipient side for one effective
// event.  Returns (0, sideNone) for non-scoring event types.
func (e *ScoreEngine) pointsForEvent(evt *db.MatchEvent, m *db.Match) (int, side) {
	switch evt.EventType {

	case db.MatchEventTypeRaidSuccessful:
		return payloadPoints(evt.Payload), participantSide(evt, m)

	case db.MatchEventTypeBonusPointAwarded:
		return 1, participantSide(evt, m)

	case db.MatchEventTypeTackleSuccessful:
		return 1, participantSide(evt, m)

	case db.MatchEventTypeSuperTackle:
		return 2, participantSide(evt, m)

	case db.MatchEventTypeAllOut:
		return allOutScore(evt.Payload, m)

	case db.MatchEventTypePenaltyAwarded:
		return payloadPoints(evt.Payload), participantSide(evt, m)

	// super_raid is an analytics/classification label only.
	// The raid's actual point contribution is already recorded via
	// the accompanying raid_successful event.  Scoring super_raid
	// separately would double-count those points.
	case db.MatchEventTypeSuperRaid:
		return 0, sideNone

	default:
		return 0, sideNone
	}
}

File: frontend/src/lib/scoring/engine.ts
import type { Match } from "@/types/api/matches"
import type { MatchEvent } from "@/types/api/match-events"

/**
 * PlayArena live-scoring engine — TypeScript port of the authoritative Go
 * engine (`backend/internal/scoring`).
 *
 * INTEGRITY CONTRACT: this fold must produce the SAME score the backend derives
 * from the same event history (`GET /matches/{id}/score`). The backend is the
 * source of truth; this engine exists for instant, offline-capable display and
 * is verified against the backend rules by golden-vector tests. Any divergence
 * is a bug in this file, never a reason to "fix" the displayed number.
 *
 * Kabaddi scoring rules (mirrors backend `scoring/rules.go`):
 *
 *   raid_successful      payload.points (>0 int)  → raiding team (event.team_id)
 *   bonus_point_awarded  +1                       → raiding team
 *   tackle_successful    +1                       → defending team
 *   super_tackle         +2                       → defending team
 *   all_out              payload.bonus_points     → OPPONENT of payload.team_id (eliminated)
 *   penalty_awarded      payload.points (>0 int)  → attributed team
 *   super_raid           0  (label only; the raid_successful carries the points)
 *   score_correction     0  (carries cancels_event_id; cancels its target)
 *   all other types      0
 *
 * EFFECTIVE SET (mirrors `GetEffectiveMatchEventsForScore`): an event is
 * cancelled — and contributes nothing — when its id appears as some event's
 * `cancels_event_id`. The score_correction events themselves remain in the set
 * but score 0. Pure, deterministic, no side effects.
 */

export type ScoringMatch = Pick<
  Match,
  "home_team_id" | "away_team_id" | "home_player_id" | "away_player_id"
>

export type Side = "home" | "away" | "none"

export interface ComputedScore {
  home: number
  away: number
}

export interface ScoreContribution {
  points: number
  side: Side
}

// Team-format matches attribute by team_id; individual-format by player_id.
// Mirrors the Go engine keying on `match.HomeTeamID.Valid`.
function isTeamMode(m: ScoringMatch): boolean {
  return m.home_team_id != null
}

// Resolves which side an event is attributed to via its participant id.
function participantSide(evt: MatchEvent, m: ScoringMatch): Side {
  if (isTeamMode(m)) {
    const id = evt.team_id
    if (!id) return "none"
    if (id === m.home_team_id) return "home"
    if (id === m.away_team_id) return "away"
    return "none"
  }
  const id = evt.player_id
  if (!id) return "none"
  if (id === m.home_player_id) return "home"
  if (id === m.away_player_id) return "away"
  return "none"
}

// Resolves a raw participant id (from an all_out payload) to a side, checking
// both team and player slots — mirrors the Go `sideByID`.
function sideById(id: string, m: ScoringMatch): Side {
  if (!id) return "none"
  if (m.home_team_id && id === m.home_team_id) return "home"
  if (m.away_team_id && id === m.away_team_id) return "away"
  if (m.home_player_id && id === m.home_player_id) return "home"
  if (m.away_player_id && id === m.away_player_id) return "away"
  return "none"
}

// Extracts payload.points as a positive integer; 0 otherwise. Mirrors the Go
// `payloadPoints` (which rejects missing / non-positive / non-integer values).
function payloadPoints(payload: Record<string, unknown> | null | undefined): number {
  const p = payload?.points
  if (typeof p === "number" && Number.isInteger(p) && p > 0) return p
  return 0
}

// all_out: payload.team_id is the ELIMINATED side; the opponent receives
// payload.bonus_points. Returns (0, none) on any payload anomaly — mirrors Go
// `allOutScore`.
function allOutScore(
  payload: Record<string, unknown> | null | undefined,
  m: ScoringMatch,
): ScoreContribution {
  const teamId = payload?.team_id
  const bonus = payload?.bonus_points
  if (typeof teamId !== "string" || teamId === "") return { points: 0, side: "none" }
  if (typeof bonus !== "number" || !Number.isInteger(bonus) || bonus <= 0) {
    return { points: 0, side: "none" }
  }
  switch (sideById(teamId, m)) {
    case "home":
      return { points: bonus, side: "away" }
    case "away":
      return { points: bonus, side: "home" }
    default:
      return { points: 0, side: "none" }
  }
}

/**
 * Point value and recipient side for a single event. Non-scoring types
 * (lifecycle, super_raid, score_correction, player state, raid_attempt/empty)
 * return 0. Exported for the read-only timeline so per-event deltas use the
 * exact same rules as the score fold.
 */
export function scoreContribution(evt: MatchEvent, m: ScoringMatch): ScoreContribution {
  switch (evt.event_type) {
    case "raid_successful":
      return { points: payloadPoints(evt.payload), side: participantSide(evt, m) }
    case "bonus_point_awarded":
      return { points: 1, side: participantSide(evt, m) }
    case "tackle_successful":
      return { points: 1, side: participantSide(evt, m) }
    case "super_tackle":
      return { points: 2, side: participantSide(evt, m) }
    case "all_out":
      return allOutScore(evt.payload, m)
    case "penalty_awarded":
      return { points: payloadPoints(evt.payload), side: participantSide(evt, m) }
    default:
      return { points: 0, side: "none" }
  }
}

/**
 * The set of event ids cancelled by a score_correction (their id appears as
 * some event's cancels_event_id). Exported so the timeline can mark cancelled
 * events without recomputing.
 */
export function cancelledEventIds(events: MatchEvent[]): Set<string> {
  const cancelled = new Set<string>()
  for (const e of events) {
    if (e.cancels_event_id) cancelled.add(e.cancels_event_id)
  }
  return cancelled
}

/** Events that still count toward the score (cancelled targets removed). */
export function effectiveEvents(events: MatchEvent[]): MatchEvent[] {
  const cancelled = cancelledEventIds(events)
  return events.filter((e) => !cancelled.has(e.id))
}

/**
 * Deterministic score fold over the effective event set. Order-independent for
 * point sums; cancellation is resolved by id, not position — so a raw log in
 * any order yields the same result the backend computes.
 */
export function computeScore(m: ScoringMatch, events: MatchEvent[]): ComputedScore {
  let home = 0
  let away = 0
  for (const e of effectiveEvents(events)) {
    const { points, side } = scoreContribution(e, m)
    if (side === "home") home += points
    else if (side === "away") away += points
  }
  return { home, away }
}
